File: l33tcode/number-of-connected-components-in-an-undirected-graph.py
from typing import List


class UnionFind:
    def __init__(self, vertices: int) -> None:
        self._tree = list(range(vertices))
        self._count = [0] * vertices

    def find(self, vertex: int) -> int:
        start = vertex
        root = vertex

        while root != self._tree[root]:
            root = self._tree[root]

        while start != root:
            start, self._tree[start] = self._tree[start], root

        return root

    def union(self, root_left: int, root_right: int) -> None:
        root_less, root_more = list(
            sorted([root_left, root_right], key=lambda x: self._count[x])
        )

        self._tree[root_less] = root_more
        self._count[root_more] += self._count[root_less]


class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        count = n

        union_find = UnionFind(n)

        for src, dst in edges:
            root_left, root_right = union_find.find(src), union_find.find(dst)

            if root_left != root_right:
                count -= 1
                union_find.union(root_left, root_right)

        return count

File: l33tcode/number-of-connected-components-in-an-undirected-graph.go
package main

type UnionFind struct {
	counts []int
	roots  []int
}

func (set *UnionFind) init() {
	for pos, _ := range set.roots {
		set.roots[pos] = pos
		set.counts[pos] = 1
	}
}

func (set *UnionFind) union(rootLeft, rootRight int) {
	rootLess, rootMore := rootLeft, rootRight

	if set.counts[rootLeft] > set.counts[rootRight] {
		rootLess, rootMore = rootMore, rootLess
	}

	set.roots[rootLess] = set.roots[rootMore]
	set.counts[rootMore] += set.counts[rootLess]
}

func (set *UnionFind) find(node int) int {
	if set.roots[node] == node {
		return node
	}

	set.roots[node] = set.find(set.roots[node])

	return set.roots[node]
}

func countComponents(n int, edges [][]int) int {
	set := UnionFind{make([]int, n), make([]int, n)}
	set.init()

	components := n

	for _, edge := range edges {
		src, dst := edge[0], edge[1]

		rootSrc, rootDst := set.find(src), set.find(dst)

		if rootSrc != rootDst {
			set.union(rootSrc, rootDst)
			components -= 1
		}

	}

	return components
}
